File: mira-hub/src/lib/workflow.ts
/**
 * runWorkflow / WorkflowRun — the Hub-side durable-workflow run tracker.
 *
 * The TypeScript twin of `mira-bots/shared/workflow.py`, writing the same
 * `workflow_runs` rows (migration 044). Use it to turn a fire-and-forget Hub
 * pipeline (ingest, KG build, CMMS sync) into a durable, observable run:
 *
 *   await runWorkflow(
 *     { workflowName: "pdf_ingest", tenantId, input: { file } },
 *     async (run) => {
 *       const text   = await run.step("parse",  () => parsePdf(file));
 *       const chunks = await run.step("chunk",  () => chunkText(text));
 *       await run.step("store", () => storeChunks(chunks));
 *       run.setOutput({ chunks: chunks.length });
 *     },
 *   );
 *   // clean return -> status "ok" (or "degraded" if a tolerated step failed)
 *   // throw        -> status "failed", error_detail captured, then re-thrown
 *
 * Contract (mirrors the Python wrapper):
 * - **Fail-open. ALWAYS.** A run-record write failure (NeonDB down, env unset,
 *   schema drift) must never break the wrapped work — including row creation. A
 *   failed create degrades to "not recorded"; the body still runs.
 * - **Steps buffered in memory, flushed ONCE at the end** — never a DB round-trip
 *   per step (Phase 2 wraps hot paths).
 * - **Pure state logic** (`computeFinalStatus` / `buildStepRecord`) is unit-tested
 *   without a DB; DB access goes through an injectable `WorkflowStore`.
 *
 * What `idempotencyKey` buys (and what it does NOT): a non-NULL key dedups the
 * `workflow_runs` row (ON CONFLICT DO NOTHING); a re-run reuses the row, bumps
 * retry_count, and sets `run.alreadySucceeded`. It does NOT auto-skip the body,
 * and it does NOT make a surface's own data-table writes idempotent — that is
 * fixed at each surface's INSERT.
 */

// The pg pool is imported lazily inside the default store (below) so this module
// can be imported for its pure helpers without instantiating a DB connection —
// the framework-free unit tests rely on this.

export type WorkflowStatus = "running" | "ok" | "degraded" | "failed";

const MAX_JSON_BYTES = 16_000;

export interface StepRecord {
  step_name: string;
  status: "ok" | "failed" | string;
  started_at: string;
  finished_at: string;
  duration_ms: number;
  artifact?: unknown;
  error?: string;
}

// ---------------------------------------------------------------------------
// Pure helpers (unit-tested without a DB)
// ---------------------------------------------------------------------------

/** Final run status from the exit condition. Pure. */
export function computeFinalStatus(threw: boolean, degraded: boolean): WorkflowStatus {
  if (threw) return "failed";
  return degraded ? "degraded" : "ok";
}

/** JSON-encode a value, bounding its size. `undefined`/`null` → null. */
export function boundedJson(value: unknown): string | null {
  if (value === undefined || value === null) return null;
  let encoded: string;
  try {
    encoded = JSON.stringify(value);
  } catch {
    encoded = JSON.stringify({ _unserializable: String(value).slice(0, MAX_JSON_BYTES) });
  }
  if (encoded === undefined) return null; // JSON.stringify(fn) etc.
  if (encoded.length > MAX_JSON_BYTES) {
    return JSON.stringify({ _truncated: true, preview: encoded.slice(0, MAX_JSON_BYTES) });
  }
  return encoded;
}

/** Assemble one step_artifacts entry. Pure. */
export function buildStepRecord(args: {
  stepName: string;
  status: string;
  startedAt: Date;
  finishedAt: Date;
  artifact?: unknown;
  error?: string;
}): StepRecord {
  const rec: StepRecord = {
    step_name: args.stepName,
    status: args.status,
    started_at: args.startedAt.toISOString(),
    finished_at: args.finishedAt.toISOString(),
    duration_ms: Math.max(0, args.finishedAt.getTime() - args.startedAt.getTime()),
  };
  if (args.artifact !== undefined && args.artifact !== null) {
    const bounded = boundedJson(args.artifact);
    rec.artifact = bounded === null ? null : JSON.parse(bounded);
  }
  if (args.error !== undefined) rec.error = args.error.slice(0, 2000);
  return rec;
}

// ---------------------------------------------------------------------------
// Store abstraction (injectable so the lifecycle is testable without pg)
// ---------------------------------------------------------------------------

export interface CreateRowInput {
  runId: string;
  workflowName: string;
  workflowVersion: string;
  tenantId: string | null;
  input: string | null; // bounded JSON string or null
  idempotencyKey: string | null;
  retryCount: number;
}

export interface CreateRowResult {
  runId: string;
  alreadySucceeded: boolean;
  retryCount: number;
}

export interface FinalizeInput {
  runId: string;
  status: WorkflowStatus;
  output: string | null;
  errorDetail: string | null;
  stepArtifacts: string; // JSON array string
}

export interface WorkflowStore {
  create(row: CreateRowInput): Promise<CreateRowResult | null>;
  finalize(input: FinalizeInput): Promise<void>;
}

async function getPool() {
  return (await import("./db")).default;
}

/** Default store — writes to NeonDB via the shared pg pool. */
export const pgWorkflowStore: WorkflowStore = {
  async create(row) {
    const pool = await getPool();
    const inserted = await pool.query(
      `INSERT INTO workflow_runs
         (run_id, workflow_name, workflow_version, tenant_id,
          status, input, idempotency_key, retry_count)
       VALUES ($1, $2, $3, $4, 'running', $5::jsonb, $6, $7)
       ON CONFLICT (idempotency_key) WHERE idempotency_key IS NOT NULL
         DO NOTHING
       RETURNING run_id`,
      [
        row.runId,
        row.workflowName,
        row.workflowVersion,
        row.tenantId,
        row.input,
        row.idempotencyKey,
        row.retryCount,
      ],
    );
    if (inserted.rows.length > 0) {
      return { runId: inserted.rows[0].run_id as string, alreadySucceeded: false, retryCount: row.retryCount };
    }

    // Conflict: reuse the existing row, surface its prior outcome, bump retry.
    const existing = await pool.query(
      `SELECT run_id, status FROM workflow_runs WHERE idempotency_key = $1`,
      [row.idempotencyKey],
    );
    if (existing.rows.length === 0) return null; // raced away → degrade to in-memory
    const runId = existing.rows[0].run_id as string;
    const alreadySucceeded = existing.rows[0].status === "ok";
    const updated = await pool.query(
      `UPDATE workflow_runs
          SET retry_count = retry_count + 1,
              status = 'running',
              started_at = NOW(),
              finished_at = NULL
        WHERE run_id = $1
       RETURNING retry_count`,
      [runId],
    );
    return {
      runId,
      alreadySucceeded,
      retryCount: updated.rows.length > 0 ? Number(updated.rows[0].retry_count) : row.retryCount,
    };
  },

  async finalize(input) {
    const pool = await getPool();
    await pool.query(
      `UPDATE workflow_runs
          SET status = $2,
              output = $3::jsonb,
              error_detail = $4,
              step_artifacts = $5::jsonb,
              finished_at = NOW()
        WHERE run_id = $1`,
      [input.runId, input.status, input.output, input.errorDetail, input.stepArtifacts],
    );
  },
};

// ---------------------------------------------------------------------------
// WorkflowRun
// ---------------------------------------------------------------------------

export interface WorkflowRunOptions {
  workflowName: string;
  version?: string;
  tenantId?: string | null;
  input?: unknown;
  idempotencyKey?: string | null;
  retryCount?: number;
  store?: WorkflowStore;
  logger?: Pick<Console, "info" | "warn" | "error">;
}

/** A JSON-able artifact summary — deliberately not `unknown` so the callback
 *  form keeps its parameter type (a union containing `unknown` collapses). */
export type ArtifactValue =
  | string
  | number
  | boolean
  | null
  | { [key: string]: unknown }
  | unknown[];

export interface StepOptions<R> {
  tolerate?: boolean;
  artifact?: ArtifactValue | ((result: R) => ArtifactValue);
}

function randomUuid(): string {
  // crypto.randomUUID is available in Node 18+ and Bun.
  return globalThis.crypto?.randomUUID?.() ?? `wfr-${Date.now()}-${Math.round(Math.random() * 1e9)}`;
}

export class WorkflowRun {
  readonly workflowName: string;
  readonly version: string;
  readonly tenantId: string | null;
  runId: string;
  status: WorkflowStatus = "running";
  output: unknown = null;
  errorDetail: string | null = null;
  alreadySucceeded = false;
  retryCount: number;

  private readonly input: unknown;
  private readonly idempotencyKey: string | null;
  private readonly store: WorkflowStore;
  private readonly log: Pick<Console, "info" | "warn" | "error">;
  private readonly steps: StepRecord[] = [];
  private degraded = false;
  private recorded = false;

  constructor(opts: WorkflowRunOptions) {
    this.workflowName = opts.workflowName;
    this.version = opts.version ?? "1.0.0";
    this.tenantId = opts.tenantId ?? null;
    this.input = opts.input ?? null;
    this.idempotencyKey = opts.idempotencyKey ?? null;
    this.retryCount = opts.retryCount ?? 0;
    this.store = opts.store ?? pgWorkflowStore;
    this.log = opts.logger ?? console;
    this.runId = randomUuid();
  }

  setOutput(value: unknown): void {
    this.output = value;
  }

  /** Run `fn` as a recorded step; return its result. See module docstring.
   *  Not tolerated → resolves to `R` (and re-throws on failure); `tolerate:true`
   *  → resolves to `R | null` (null when the step failed and was swallowed). */
  step<R>(name: string, fn: () => Promise<R> | R, opts?: Omit<StepOptions<R>, "tolerate"> & { tolerate?: false }): Promise<R>;
  step<R>(name: string, fn: () => Promise<R> | R, opts: Omit<StepOptions<R>, "tolerate"> & { tolerate: true }): Promise<R | null>;
  async step<R>(name: string, fn: () => Promise<R> | R, opts: StepOptions<R> = {}): Promise<R | null> {
    const started = new Date();
    try {
      const result = await fn();
      const artifact =
        typeof opts.artifact === "function"
          ? (opts.artifact as (r: R) => ArtifactValue)(result)
          : opts.artifact;
      this.steps.push(
        buildStepRecord({ stepName: name, status: "ok", startedAt: started, finishedAt: new Date(), artifact }),
      );
      return result;
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      this.steps.push(
        buildStepRecord({ stepName: name, status: "failed", startedAt: started, finishedAt: new Date(), error: message }),
      );
      this.log.warn(`workflow ${this.workflowName} step ${name} failed: ${message}`);
      if (opts.tolerate) {
        this.degraded = true;
        return null;
      }
      throw err;
    }
  }

  /** Record a step that doesn't fit the `step(fn)` call-wrap shape. */
  recordStep(
    name: string,
    status: string,
    opts: { artifact?: unknown; error?: string; startedAt?: Date; finishedAt?: Date } = {},
  ): void {
    const now = new Date();
    if (status === "failed") this.degraded = true;
    this.steps.push(
      buildStepRecord({
        stepName: name,
        status,
        startedAt: opts.startedAt ?? now,
        finishedAt: opts.finishedAt ?? now,
        artifact: opts.artifact,
        error: opts.error,
      }),
    );
  }

  /** @internal — called by runWorkflow. */
  async _begin(): Promise<void> {
    try {
      const res = await this.store.create({
        runId: this.runId,
        workflowName: this.workflowName,
        workflowVersion: this.version,
        tenantId: this.tenantId,
        input: boundedJson(this.input),
        idempotencyKey: this.idempotencyKey,
        retryCount: this.retryCount,
      });
      if (res) {
        this.runId = res.runId;
        this.alreadySucceeded = res.alreadySucceeded;
        this.retryCount = res.retryCount;
        this.recorded = true;
      }
    } catch (err) {
      this.log.warn(`workflow run create skipped (degraded to in-memory): ${String(err)}`);
    }
    this.log.info(
      `workflow ${this.workflowName} v${this.version} started (run_id=${this.runId} tenant=${this.tenantId})`,
    );
  }

  /** @internal — called by runWorkflow. */
  async _finish(threw: boolean, error?: unknown): Promise<void> {
    this.status = computeFinalStatus(threw, this.degraded);
    if (threw && this.errorDetail === null) {
      this.errorDetail = error instanceof Error ? `${error.name}: ${error.message}` : String(error);
    }
    if (this.recorded) {
      try {
        await this.store.finalize({
          runId: this.runId,
          status: this.status,
          output: boundedJson(this.output),
          errorDetail: this.errorDetail,
          stepArtifacts: JSON.stringify(this.steps),
        });
      } catch (err) {
        this.log.warn(`workflow run finalize skipped: ${String(err)}`);
      }
    }
    const line = `workflow ${this.workflowName} finished status=${this.status} (run_id=${this.runId} steps=${this.steps.length})`;
    if (this.status === "failed") this.log.error(line);
    else this.log.info(line);
  }

  /** Test-only accessor for the buffered step records. */
  get _steps(): readonly StepRecord[] {
    return this.steps;
  }
}

/**
 * Run `fn` inside a durable workflow envelope. Creates the run record, runs the
 * body, finalizes status (ok/degraded/failed), and re-throws any error after
 * recording it. Returns whatever `fn` returns.
 */
export async function runWorkflow<T>(
  opts: WorkflowRunOptions,
  fn: (run: WorkflowRun) => Promise<T>,
): Promise<T> {
  const run = new WorkflowRun(opts);
  await run._begin();
  try {
    const result = await fn(run);
    await run._finish(false);
    return result;
  } catch (err) {
    await run._finish(true, err);
    throw err;
  }
}

File: mira-bots/shared/workflow.py
"""WorkflowRun — the shared durable-workflow run tracker (audit primitive #1).

Every major MIRA action should be *a machine with a dashboard, not a hidden code
path*. This wrapper turns any call-chain into a durable, observable run by
writing one `workflow_runs` row (Hub migration 044) at start, recording each
step in memory, and flushing the final status + step artifacts at exit.

Usage::

    async with WorkflowRun("pdf_ingest", tenant_id=tid, input={"file": name}) as run:
        text   = await run.step("parse",  parse_pdf, file)
        chunks = await run.step("chunk",  chunk_text, text)
        await run.step("store", store_chunks, chunks)
        run.set_output({"chunks": len(chunks)})
    # clean exit  -> status="ok" (or "degraded" if a tolerated step failed)
    # exception   -> status="failed", error_detail captured, then re-raised

Design constraints (mirror decision_trace.py — the established precedent):

- **Fail-open. ALWAYS.** A run-record write failure (NeonDB down, env unset,
  schema drift) must NEVER block, delay, or fail the wrapped work. This includes
  ``__aenter__`` — if the create-row INSERT fails, the body still runs; the run
  simply degrades to "not recorded". The wrapper is observational, not
  load-bearing.
- **Event loop never blocked.** DB writes are offloaded to a worker thread via
  ``run_in_executor`` with a hard timeout. Steps are buffered in memory and
  flushed ONCE at exit — never a DB round-trip per ``step()`` (Phase 2 wraps hot
  paths like ``engine.handle_message``).
- **Lazy imports.** sqlalchemy is imported inside the worker so services without
  it still boot and import this module.
- **Pure state logic.** ``compute_final_status`` / ``build_step_record`` are pure
  functions — the lifecycle is unit-tested without a live NeonDB.

What idempotency_key buys (and what it does NOT):
- A non-NULL key dedups the `workflow_runs` row (ON CONFLICT DO NOTHING on the
  partial unique index); a re-run reuses the existing run_id and bumps
  retry_count instead of inserting a second row.
- ``run.already_succeeded`` lets the caller branch ("skip the work").
- It does **NOT** auto-skip the with-body (a context manager always runs its
  body), and it does **NOT** make a surface's own data-table writes idempotent —
  that is fixed at each surface's INSERT (the audit's anti-pattern #5), not here.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import uuid
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable, Optional, Union

logger = logging.getLogger("mira.workflow")

# DB writes are bounded so a slow/unreachable NeonDB never stalls the wrapped
# work. Same 2s budget decision_trace uses for its observational insert.
_TIMEOUT_SECONDS = 2

# Cap any single artifact / output payload so a runaway dict can't bloat the row
# or blow the JSONB. Truncation is recorded so it is never silently lossy.
_MAX_JSON_BYTES = 16_000

_VALID_STATUS = ("running", "ok", "degraded", "failed")


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _connect_args(url: str) -> dict[str, Any]:
    """SSL connect args for create_engine.

    Neon requires ``sslmode=require``; we default to it (matching
    decision_trace.py) UNLESS the URL already carries an explicit sslmode — that
    lets a local/throwaway Postgres pass ``?sslmode=disable`` for integration
    tests without the wrapper forcing TLS it doesn't have.
    """
    if "sslmode" in url:
        return {}
    return {"sslmode": "require"}


# ---------------------------------------------------------------------------
# Pure helpers (unit-tested without a DB)
# ---------------------------------------------------------------------------


def compute_final_status(exc_type: Optional[type], degraded: bool) -> str:
    """Final run status from the exit condition. Pure.

    Exception during the with-body -> "failed". Clean exit -> "degraded" if any
    tolerated step failed, else "ok".
    """
    status = "failed" if exc_type is not None else ("degraded" if degraded else "ok")
    assert status in _VALID_STATUS  # postcondition: only ever a valid enum value
    return status


def _bounded_json(value: Any) -> Optional[str]:
    """JSON-encode a value, bounding its size. None passes through as None."""
    if value is None:
        return None
    try:
        encoded = json.dumps(value, default=str)
    except (TypeError, ValueError):
        encoded = json.dumps({"_unserializable": str(value)[:_MAX_JSON_BYTES]})
    if len(encoded) > _MAX_JSON_BYTES:
        return json.dumps({"_truncated": True, "preview": encoded[:_MAX_JSON_BYTES]})
    return encoded


def build_step_record(
    *,
    step_name: str,
    status: str,
    started_at: datetime,
    finished_at: datetime,
    artifact: Any = None,
    error: Optional[str] = None,
) -> dict[str, Any]:
    """Assemble one step_artifacts entry. Pure.

    ``artifact`` is stored as a bounded JSON-able summary; large/unserialisable
    values are truncated, never dropped silently.
    """
    rec: dict[str, Any] = {
        "step_name": step_name,
        "status": status,
        "started_at": started_at.isoformat(),
        "finished_at": finished_at.isoformat(),
        "duration_ms": int((finished_at - started_at).total_seconds() * 1000),
    }
    if artifact is not None:
        bounded = _bounded_json(artifact)
        # store the parsed form so step_artifacts stays a JSON array, not strings
        rec["artifact"] = json.loads(bounded) if bounded is not None else None
    if error is not None:
        rec["error"] = error[:2000]
    return rec


# ---------------------------------------------------------------------------
# WorkflowRun
# ---------------------------------------------------------------------------


class WorkflowRun:
    """Durable workflow run tracker. See module docstring for usage."""

    def __init__(
        self,
        workflow_name: str,
        *,
        version: str = "1.0.0",
        tenant_id: Optional[str] = None,
        input: Optional[dict] = None,
        idempotency_key: Optional[str] = None,
        retry_count: int = 0,
        db_url: Optional[str] = None,
        log: Optional[logging.Logger] = None,
    ) -> None:
        self.workflow_name = workflow_name
        self.version = version
        self.tenant_id = tenant_id
        self.input = input
        self.idempotency_key = idempotency_key
        self.retry_count = retry_count
        self._db_url = db_url  # None -> resolved from env at write time
        self._log = log or logger

        # Local run_id so in-memory tracking works even when the DB is absent;
        # replaced by the DB's run_id if a row is created/reused.
        self.run_id: str = str(uuid.uuid4())
        self.status: str = "running"
        self.output: Any = None
        self.error_detail: Optional[str] = None
        self.already_succeeded: bool = False
        self._steps: list[dict[str, Any]] = []
        self._degraded: bool = False
        self._recorded: bool = False  # True once a DB row exists for run_id

    # -- public API --------------------------------------------------------

    def set_output(self, value: Any) -> None:
        """Set the run's output payload (flushed at exit)."""
        self.output = value

    async def step(
        self,
        name: str,
        fn: Callable[..., Union[Any, Awaitable[Any]]],
        *args: Any,
        tolerate: bool = False,
        artifact: Optional[Union[Any, Callable[[Any], Any]]] = None,
        **kwargs: Any,
    ) -> Any:
        """Run ``fn(*args, **kwargs)`` as a recorded step; return its result.

        ``fn`` may be sync or async. On success the step is recorded "ok" and the
        result returned. On exception the step is recorded "failed":
        - ``tolerate=False`` (default): re-raise -> run ends "failed".
        - ``tolerate=True``: swallow, mark the run "degraded", return None.

        ``artifact`` is a JSON-able summary, or a callable taking the result and
        returning one. Keep it small — it is bounded to ~16 KB.
        """
        started = _utcnow()
        try:
            result = fn(*args, **kwargs)
            if asyncio.iscoroutine(result) or asyncio.isfuture(result):
                result = await result
        except Exception as exc:  # noqa: BLE001 — record then propagate per tolerate
            self._steps.append(
                build_step_record(
                    step_name=name,
                    status="failed",
                    started_at=started,
                    finished_at=_utcnow(),
                    error=str(exc),
                )
            )
            self._log.warning("workflow %s step %s failed: %s", self.workflow_name, name, exc)
            if tolerate:
                self._degraded = True
                return None
            raise

        resolved_artifact = artifact(result) if callable(artifact) else artifact
        self._steps.append(
            build_step_record(
                step_name=name,
                status="ok",
                started_at=started,
                finished_at=_utcnow(),
                artifact=resolved_artifact,
            )
        )
        return result

    def record_step(
        self,
        name: str,
        status: str,
        *,
        artifact: Any = None,
        error: Optional[str] = None,
        started_at: Optional[datetime] = None,
        finished_at: Optional[datetime] = None,
    ) -> None:
        """Record a step that doesn't fit the ``step(fn)`` call-wrap shape.

        ``status="failed"`` marks the run degraded (the caller chose to continue).
        """
        now = _utcnow()
        if status == "failed":
            self._degraded = True
        self._steps.append(
            build_step_record(
                step_name=name,
                status=status,
                started_at=started_at or now,
                finished_at=finished_at or now,
                artifact=artifact,
                error=error,
            )
        )

    # -- context manager ---------------------------------------------------

    async def __aenter__(self) -> "WorkflowRun":
        await self._create_row()
        self._log.info(
            "workflow %s v%s started (run_id=%s tenant=%s)",
            self.workflow_name,
            self.version,
            self.run_id,
            self.tenant_id,
        )
        return self

    async def __aexit__(self, exc_type, exc, _tb) -> bool:  # noqa: ANN001
        self.status = compute_final_status(exc_type, self._degraded)
        if exc_type is not None and self.error_detail is None:
            self.error_detail = f"{getattr(exc_type, '__name__', exc_type)}: {exc}"
        await self._finalize()
        level = logging.ERROR if self.status == "failed" else logging.INFO
        self._log.log(
            level,
            "workflow %s finished status=%s (run_id=%s steps=%d)",
            self.workflow_name,
            self.status,
            self.run_id,
            len(self._steps),
        )
        return False  # never suppress the wrapped exception

    # -- DB writes (fail-open, executor-offloaded, bounded) ----------------

    def _url(self) -> Optional[str]:
        return self._db_url or os.environ.get("NEON_DATABASE_URL")

    async def _create_row(self) -> None:
        """Insert (or reuse, on idempotency conflict) the run row. Fail-open."""
        if not self._url():
            return  # run-record storage disabled — purely in-memory
        try:
            await asyncio.wait_for(
                asyncio.get_running_loop().run_in_executor(None, self._create_row_sync),
                timeout=_TIMEOUT_SECONDS,
            )
        except Exception as exc:  # noqa: BLE001
            self._log.warning("workflow run create skipped (degraded to in-memory): %s", exc)

    def _finalize(self) -> Awaitable[None]:
        async def _run() -> None:
            if not self._recorded or not self._url():
                return  # no row to update (never created) — nothing to flush
            try:
                await asyncio.wait_for(
                    asyncio.get_running_loop().run_in_executor(None, self._finalize_sync),
                    timeout=_TIMEOUT_SECONDS,
                )
            except Exception as exc:  # noqa: BLE001
                self._log.warning("workflow run finalize skipped: %s", exc)

        return _run()

    def _create_row_sync(self) -> None:
        url = self._url()
        if not url:
            return
        from sqlalchemy import create_engine
        from sqlalchemy import text as sql_text
        from sqlalchemy.pool import NullPool

        engine = create_engine(
            url,
            poolclass=NullPool,
            connect_args=_connect_args(url),
            pool_pre_ping=True,
        )
        params = {
            "run_id": self.run_id,
            "workflow_name": self.workflow_name,
            "workflow_version": self.version,
            "tenant_id": self.tenant_id,
            "input": _bounded_json(self.input),
            "idempotency_key": self.idempotency_key,
            "retry_count": self.retry_count,
        }
        with engine.begin() as conn:
            row = conn.execute(
                sql_text(
                    """
                    INSERT INTO workflow_runs
                      (run_id, workflow_name, workflow_version, tenant_id,
                       status, input, idempotency_key, retry_count)
                    VALUES
                      (CAST(:run_id AS UUID), :workflow_name, :workflow_version,
                       :tenant_id, 'running', CAST(:input AS JSONB),
                       :idempotency_key, :retry_count)
                    ON CONFLICT (idempotency_key)
                      WHERE idempotency_key IS NOT NULL
                      DO NOTHING
                    RETURNING run_id
                    """
                ),
                params,
            ).fetchone()

            if row is not None:
                self.run_id = str(row[0])
                self._recorded = True
                return

            # Conflict: a row with this idempotency_key already exists. Read its
            # prior outcome (to surface already_succeeded), then reuse the row and
            # bump retry_count — one logical run, re-executed, never a 2nd row.
            existing = conn.execute(
                sql_text("SELECT run_id, status FROM workflow_runs WHERE idempotency_key = :k"),
                {"k": self.idempotency_key},
            ).fetchone()
            if existing is None:
                return  # raced away; degrade to in-memory
            self.run_id = str(existing[0])
            self.already_succeeded = existing[1] == "ok"
            updated = conn.execute(
                sql_text(
                    """
                    UPDATE workflow_runs
                       SET retry_count = retry_count + 1,
                           status = 'running',
                           started_at = NOW(),
                           finished_at = NULL
                     WHERE run_id = CAST(:run_id AS UUID)
                    RETURNING retry_count
                    """
                ),
                {"run_id": self.run_id},
            ).fetchone()
            if updated is not None:
                self.retry_count = int(updated[0])
            self._recorded = True

    def _finalize_sync(self) -> None:
        url = self._url()
        if not url:
            return
        from sqlalchemy import create_engine
        from sqlalchemy import text as sql_text
        from sqlalchemy.pool import NullPool

        engine = create_engine(
            url,
            poolclass=NullPool,
            connect_args=_connect_args(url),
            pool_pre_ping=True,
        )
        with engine.begin() as conn:
            conn.execute(
                sql_text(
                    """
                    UPDATE workflow_runs
                       SET status = :status,
                           output = CAST(:output AS JSONB),
                           error_detail = :error_detail,
                           step_artifacts = CAST(:step_artifacts AS JSONB),
                           finished_at = NOW()
                     WHERE run_id = CAST(:run_id AS UUID)
                    """
                ),
                {
                    "run_id": self.run_id,
                    "status": self.status,
                    "output": _bounded_json(self.output),
                    "error_detail": (self.error_detail or None),
                    "step_artifacts": json.dumps(self._steps, default=str),
                },
            )
